src/ledger/user.js

```diff
@@ -0,0 +1,105 @@
+// @flow
+
+/**
+ * This module has a core data type identifying SourceCred users.
+ *
+ * The scope for this data type is to model:
+ * - a unique identifier for each user
+ * - a unique (renameable) username they choose
+ * - the address of every node they correspond to in the graph
+ *
+ * Unlike most other state in SourceCred, the User state is
+ * nondeterministically generated by SourceCred itself, and then persisted
+ * long-term within the instance.
+ *
+ * This is in contrast to Graph data, which usually comes from an external
+ * source, and is not persisted long-term, but instead is re-generated when
+ * needed.
+ *
+ * In particular, this kernel of user data is stored within the core ledger,
+ * since it's necessary to track consistently when tracking Grain distribution.
+ * This type should not grow to include all the data that the UI will
+ * eventually want to show; that should be kept in a different data store which
+ * isn't being used as a transaction ledger.
+ *
+ */
+import {
+  type Uuid,
+  parser as uuidParser,
+  random as randomUuid,
+} from "../util/uuid";
+import * as C from "../util/combo";
+import {
+  type NodeAddressT,
+  NodeAddress,
+  type Node as GraphNode,
+} from "../core/graph";
+
+/**
+ * We validate usernames using GitHub-esque rules.
+ */
+export opaque type Username: string = string;
+const USERNAME_PATTERN = /^@?([A-Za-z0-9-_]+)$/;
```

Comment:
We should not allow underscores in usernames; they’re not valid in DNS
labels, which precludes some useful behavior (same reason that GitHub
disallows underscores).

Reply:
Done